201-front-door-premium-vm-private-link/modules/front-door.bicep

```diff
@@ -0,0 +1,167 @@
+param privateEndpointResourceId string = ''
+
+@description('If you are using Private Link to connect to the origin, this should specify the resource type of the Private Link resource. The allowed value will depend on the specific Private Link resource type you are using. If you are not using Private Link then this should be empty.')
+param privateLinkResourceType string = ''
+
+@description('If you are using Private Link to connect to the origin, this should specify the location of the Private Link resource. If you are not using Private Link then this should be empty.')
+param privateEndpointLocation string = ''
+
+// When connecting to Private Link origins, we need to assemble the privateLinkOriginDetails object with various pieces of data.
+var isPrivateLinkOrigin = (privateEndpointResourceId != '')
+var privateLinkOriginDetails = {
+  privateLink: {
+    id: privateEndpointResourceId
+  }
+  groupId: (privateLinkResourceType != '') ? privateLinkResourceType : null
+  privateLinkLocation: privateEndpointLocation
+  requestMessage: 'Please approve this connection.'
+}
+
+var profileName = 'MyFrontDoor'
+var originGroupName = 'MyOriginGroup'
+var originName = 'MyOrigin'
+var routeName = 'MyRoute'
+
+resource profile 'Microsoft.Cdn/profiles@2020-09-01' = {
+  name: profileName
+  location: 'global'
+  sku: {
+    name: skuName
+  }
+}
+
+resource endpoint 'Microsoft.Cdn/profiles/afdEndpoints@2020-09-01' = {
+  name: '${profile.name}/${endpointName}'
+  location: 'global'
+  properties: {
+    originResponseTimeoutSeconds: 240
+    enabledState: 'Enabled'
+  }
+}
+
+resource originGroup 'Microsoft.Cdn/profiles/originGroups@2020-09-01' = {
+  name: '${profile.name}/${originGroupName}'
+  properties: {
+    loadBalancingSettings: {
+      sampleSize: 4
+      successfulSamplesRequired: 3
+    }
+    healthProbeSettings: {
+      probePath: '/'
+      probeRequestType: 'HEAD'
+      probeProtocol: 'Http'
+      probeIntervalInSeconds: 100
+    }
+  }
+}
+
+resource origin 'Microsoft.Cdn/profiles/originGroups/origins@2020-09-01' = {
+  name: '${originGroup.name}/${originName}'
+  properties: {
+    hostName: originHostName
+    httpPort: 80
+    httpsPort: 443
+    originHostHeader: endpoint.properties.hostName // This probably isn't the right host header for most scenarios - you will need to look at what makes sense for your solution.
+    priority: 1
+    weight: 1000
+    sharedPrivateLinkResource: isPrivateLinkOrigin ? privateLinkOriginDetails : null
+  }
+}
+
+resource route 'Microsoft.Cdn/profiles/afdEndpoints/routes@2020-09-01' = {
+  name: '${endpoint.name}/${routeName}'
+  dependsOn: [
+    origin // This explicit dependency is required to ensure that the origin group is not empty when the route is created.

```

Comment:
did bicep not create this for you? what happens when it doesn't?  (it's an odd pattern)

Reply:
No - this is a quirk of the Front Door ARM RP. There is a clear dependency between the route and the origin group, and also between the origin group and the origin - and Bicep detects these. However, this explicit dependency is required because Front Door has a rule that a route can't be created if the origin group it references is empty. I've spoken to the PG about this as I think it's not intuitive - but that's the behaviour they need to go with for now at least.

Reply:
Thank you for trying (those conversations can be frustrating...)